polar-core/src/kb.rs

```diff
@@ -75,11 +93,379 @@ impl KnowledgeBase {
+                                &rule_instance.fields,
+                            )
+                        {
+                            RuleParamMatch::False(format!("Rule specializer {} on parameter {} did not match prototype specializer {} because the specializer fields did not match.", rule_instance.to_polar(), index, prototype_instance.to_polar()))
+                        } else {
+                            RuleParamMatch::True
+                        }
+                    } else {
+                        return Err(error::OperationalError::InvalidState(format!(
+                                "All registered classes must have a registered MRO. Class {} does not have a registered MRO.",
+                                &rule_instance.tag
+                            )).into());
+                    }
+                } else {
+                    unreachable!("Unregistered specializer classes should be caught before this point.");
+                }
+            }
+            (Pattern::Dictionary(prototype_fields), Pattern::Dictionary(rule_fields))
+            | (
+                Pattern::Dictionary(prototype_fields),
+                Pattern::Instance(InstanceLiteral {
+                    tag: _,
+                    fields: rule_fields,
+                }),
+            ) => {
+                if self.param_fields_match(prototype_fields, rule_fields) {
+                    RuleParamMatch::True
+                } else {
+                    RuleParamMatch::False(format!("Specializer mismatch on parameter {}. Rule specializer fields {:#?} do not match prototype specializer fields {:#?}.", index, rule_fields, prototype_fields))
+                }
+            }
+            (
+                Pattern::Instance(InstanceLiteral {
+                    tag,
+                    fields: prototype_fields,
+                }),
+                Pattern::Dictionary(rule_fields),
+            ) if tag == &sym!("Dictionary") => {
+                if self.param_fields_match(prototype_fields, rule_fields) {
+                    RuleParamMatch::True
+                } else {
+                    RuleParamMatch::False(format!("Specializer mismatch on parameter {}. Rule specializer fields {:#?} do not match prototype specializer fields {:#?}.", index, rule_fields, prototype_fields))
+                }
+            }
+            (_, _) => {
+                RuleParamMatch::False(format!("Mismatch on parameter {}. Rule parameter {:#?} does not match prototype parameter {:#?}.", index, prototype_pattern, rule_pattern))
+            }
+        })
+    }
+
+    /// Check that a rule parameter that is a value matches a prototype parameter that is a value
+    fn check_value_param(
+        &self,
+        index: usize,
+        rule_value: &Value,
+        prototype_value: &Value,
+    ) -> PolarResult<RuleParamMatch> {
+        Ok(match (prototype_value, rule_value) {
+            (Value::List(prototype_list), Value::List(rule_list)) => {
```

Comment:
Might want to reuse the list unification logic from the VM to account for *rest vars.

Also not totally sure w/o putting more thought into it what the semantics should be for the interactions between prototypes & *rest vars.